File: src/map/mapflag.cpp
#include "mapflag.hpp"
//    mapflag.cpp - Implementation of map flag settings
//
//    Copyright © 2014 Ben Longbons <b.r.longbons@gmail.com>
//
//    This file is part of The Mana World (Athena server)
//
//    This program is free software: you can redistribute it and/or modify
//    it under the terms of the GNU General Public License as published by
//    the Free Software Foundation, either version 3 of the License, or
//    (at your option) any later version.
//
//    This program is distributed in the hope that it will be useful,
//    but WITHOUT ANY WARRANTY; without even the implied warranty of
//    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//    GNU General Public License for more details.
//
//    You should have received a copy of the GNU General Public License
//    along with this program.  If not, see <http://www.gnu.org/licenses/>.

#include "../poison.hpp"


namespace tmwa
{
// because bitfields, that's why

bool MapFlags::get(MapFlag mf) const
{
    return flags & static_cast<unsigned>(mf);
}

void MapFlags::set(MapFlag mf, bool val)
{
    if (val)
        flags |= static_cast<unsigned>(mf);
    else
        flags &=~ static_cast<unsigned>(mf);
}

template<>
bool extract<MapFlag, void, void>(XString str, MapFlag *mf)
{
    const struct
    {
        LString str;
        MapFlag id;
    } flags[] =
    {
        //{"alias"_s, MapFlag::ALIAS},
        //{"nomemo"_s, MapFlag::NOMEMO},
        {"noteleport"_s, MapFlag::NOTELEPORT},
        {"noreturn"_s, MapFlag::NORETURN},
        {"monster_noteleport"_s, MapFlag::MONSTER_NOTELEPORT},
        {"nosave"_s, MapFlag::NOSAVE},
        //{"nobranch"_s, MapFlag::NOBRANCH},
        {"nopenalty"_s, MapFlag::NOPENALTY},
        {"pvp"_s, MapFlag::PVP},
        {"pvp_noparty"_s, MapFlag::PVP_NOPARTY},
        //{"pvp_noguild"_s, MapFlag::PVP_NOGUILD},
        //{"pvp_nightmaredrop"_s, MapFlag::PVP_NIGHTMAREDROP},
        {"pvp_nocalcrank"_s, MapFlag::PVP_NOCALCRANK},
        //{"gvg"_s, MapFlag::GVG},
        //{"gvg_noparty"_s, MapFlag::GVG_NOPARTY},
        //{"nozenypenalty"_s, MapFlag::NOZENYPENALTY},
        //{"notrade"_s, MapFlag::NOTRADE},
        //{"noskill"_s, MapFlag::NOSKILL},
        {"nowarp"_s, MapFlag::NOWARP},
        {"nowarpto"_s, MapFlag::NOWARPTO},
        {"nopvp"_s, MapFlag::NOPVP},
        //{"noicewall"_s, MapFlag::NOICEWALL},
        {"snow"_s, MapFlag::SNOW},
        {"fog"_s, MapFlag::FOG},
        {"sakura"_s, MapFlag::SAKURA},
        {"leaves"_s, MapFlag::LEAVES},
        {"rain"_s, MapFlag::RAIN},
        {"no_player_drops"_s, MapFlag::NO_PLAYER_DROPS},
        {"town"_s, MapFlag::TOWN},
        {"outside"_s, MapFlag::OUTSIDE},
        {"resave"_s, MapFlag::RESAVE},
    };
    for (auto& pair : flags)
        if (str == pair.str)
        {
            *mf = pair.id;
            return true;
        }
    return false;
}

MapFlag map_flag_from_int(int shift)
{
    return static_cast<MapFlag>(1 << shift);
}
} // namespace tmwa

File: src/map/mapflag.py
class MapFlags(object):
    __slots__ = ('_value')
    name = 'tmwa::map::MapFlags'
    enabled = True

    def __init__(self, value):
        self._value = value['flags']

    def to_string(self):
        i = int(self._value)
        s = []
        for n, v in MapFlags.junk:
            v = 1 << v
            if i & v:
                i -= v
                s.append(n)
        if i or not s:
            s.append('%#08x' % i)
        return 'MapFlags(%s)' % (' | '.join(s))

    junk = [
            #('ALIAS', 21),
            #('NOMEMO', 0),
            ('NOTELEPORT', 1),
            ('NORETURN', 22),
            ('MONSTER_NOTELEPORT', 23),
            ('NOSAVE', 2),
            #('NOBRANCH', 3),
            ('NOPENALTY', 4),
            ('PVP', 6),
            ('PVP_NOPARTY', 7),
            #('PVP_NOGUILD', 8),
            #('PVP_NIGHTMAREDROP', 24),
            ('PVP_NOCALCRANK', 25),
            #('GVG', 9),
            #('GVG_NOPARTY', 10),
            #('NOZENYPENALTY', 5),
            #('NOTRADE', 11),
            #('NOSKILL', 12),
            ('NOWARP', 13),
            ('NOWARPTO', 26),
            ('NOPVP', 14),
            ('MASK', 15),
            ('SNOW', 16),
            ('FOG', 17),
            ('SAKURA', 18),
            ('LEAVES', 19),
            ('RAIN', 20),
            ('NO_PLAYER_DROPS', 27),
            ('TOWN', 28),
            ('OUTSIDE', 29),
            ('RESAVE', 30),
    ]
    tests = [
            ('reinterpret_cast<const tmwa::map::MapFlags&>(static_cast<const unsigned int&>(0x80000000))', 'MapFlags(0x80000000)'),
            ('reinterpret_cast<const tmwa::map::MapFlags&>(static_cast<const unsigned int&>(0xf0000000))', 'MapFlags(TOWN | OUTSIDE | RESAVE | 0x80000000)'),
    ] + [
            ('tmwa::map::MapFlags(); value.set(tmwa::map::MapFlag::%s, true)' % n, 'MapFlags(%s)' % n)
            for (n, _) in junk
    ] + [
            ('reinterpret_cast<const tmwa::map::MapFlags&>(static_cast<const unsigned int&>(1 << %d))' % i, 'MapFlags(%s)' % n)
            for (n, i) in junk
    ]
